Lazy tree implementation

roysc opened the pull request:
Rewrites most of the SMT:

* current tree state exists as allocated node objects
* flushes to MapStore on `Save()`
* tree is loaded lazily, and caches nodes as they are written/read
* value storage MapStore is removed
* adds an option to set path-hashing function (identity for raw key)
* types are refactored and cleaned up

roysc committed:
Remove *ForRoot methods

roysc committed:
Remove value storage

roysc committed:
Refactor tests, decouple SMT interface

misc. cleanup

roysc committed:
Implement lazy SMT

test empty root

improve old tests

partial caching w/ stub nodes

cache node digest

roysc committed:
Configurable path hasher; refactor hasher

adlerjohn committed:
Remove changelog checker in favor of automatic on release. (#68)

roysc committed:
Improve old tests

test empty root
expand orphan tests

roysc committed:
Refactor and remove old impl

and clean up, change method sigs, nits

roysc committed:
Rename types

roysc committed:
Rename methods

roysc committed:
Fix fuzzer

roysc committed:
Publicize names

roysc committed:
Refactor hashers

keep path/node/value hashing distinct

roysc committed:
Allow arbitrary leaf value data

roysc committed:
add go.sum

roysc committed:
Add Commit/CachedMapStore

roysc committed:
small refactors, renames

roysc committed:
Get: make iterative, update cache

roysc committed:
nits

roysc committed:
in-place orphans, avoid copying

roysc committed:
improve bulk tests

squash! [wip] less stupid bulk tests

i-norden commented on cached_mapstore.go:
```diff
@@ -0,0 +1,71 @@
+
+// Set updates the value for a key.
+func (cm *CachedMapStore) Set(key []byte, value []byte) error {
+	cm.cache[string(key)] = value
```
Do we need to check for and clear this key in the `deletes` cache?

i-norden commented on cached_mapstore.go:
```diff
@@ -0,0 +1,71 @@
+
+func (cm *CachedMapStore) Commit() error {
+	for k, _ := range cm.deletes {
+		if err := cm.db.Delete([]byte(k)); err != nil {
```
Related to the comment above in `Set`, I see now we do not because we flush deletes to the MapStore first so the delete will be overwritten. But I wonder if it would be more optimal to avoid flushing a `Delete` here if it could have been removed from the cache during `Set`? I suspect overall it is not, because this access pattern happens relatively infrequently and so the additional overhead of having to check the `deletes` cache on every `Set` is not justified.

roysc replied:
Right, we don't need to clear deletions but could, as a small performance tradeoff in favor of commit.

i-norden commented on cached_mapstore.go:
```diff
@@ -0,0 +1,71 @@
+			return err
+		}
+	}
+	cm.cache = map[string]operation{}
```
Need to reset the `deletes` cache too.

roysc replied:
d'oh!

i-norden commented on options.go:
```diff
@@ -1,4 +1,12 @@
+// Option is a function that configures SparseMerkleTree.
+type Option func(*SMT)
+
+func SetPathHasher(ph PathHasher) Option {
```
Nice added generality with these two setters, will make testing changing the hashing function super simple.

i-norden commented on cached_mapstore.go:
```diff
@@ -0,0 +1,71 @@
+// CommitMapStore composes a Commit interface with a MapStore interface
+type CommitMapStore interface {
+	MapStore
+	Commit() error
```
Not sure we need this `Commit()` method any more since the base interface has the `Save()` method. And if so, I think we should rename `Save()` to `Commit()`.

roysc replied:
Actually I think we don't need the interface, but we still want `Commit` as a method, so the user has control over flushing the cache.

i-norden commented on smt.go:
```diff
@@ -1,457 +1,425 @@
+		return newLeaf, nil
+	}
+	if leaf, ok := node.(*leafNode); ok {
+		// TODO (optimization) - can just count [depth:]
```
Probably worth making this optimization as we need to squeeze every bit of performance out of this that we can 😅

i-norden commented on cached_mapstore.go:
```diff
@@ -0,0 +1,71 @@
+			return err
+		}
+	}
+	cm.cache = map[string]operation{}
```
If we can get a general idea of approx how many operations there are per block (or some lower bound estimate), I wonder if we can get some minor performance improvement by `make`ing this map with a starting capacity to minimize copy ops and memory allocations.

roysc replied:
Right, we can add a capacity field to the struct to anticipate that

i-norden commented on smt.go:
```diff
@@ -1,457 +1,425 @@
+	// Current state of tree
+	tree treeNode
+	// Hashes of persisted nodes deleted from tree
+	orphans [][]byte
```
Also wondering if we can optimize by minimizing copies and allocations by having a starting capacity here, and basically any other slice we have to continually resize.

roysc replied:
True. The expected capacity would depend on the tree size, so we can track size too.

i-norden reviewed (commented):
Amazing work, just a few comments and I'm going to continue picking through it looking for optimizations.

i-norden commented on smt.go:
```diff
@@ -1,457 +1,425 @@
+			return
+		}
+	}
+	smt.orphans = nil
```
To be clear, here is where we would `make` with a starting capacity. Or does this need to be `nil`?

roysc replied:
No, we could add that

roysc committed:
impl. extension nodes

roysc committed:
cached mapstore updates

roysc committed:
nit - rename

roysc committed:
rename save -> commit

roysc committed:
clean up

roysc committed:
count path prefix from depth

roysc committed:
fix orphan deletion bug

roysc commented:
Updated with
* extension nodes
* tweaks: avoid cloning nodes when orphaning; use array pointer instead of slice as path
* improve bulk tests
* tried pre-allocating orphan slices (based on running average tree depth) but it produced slowdowns (+5~10%) on update/delete, so left it out :\
* fixed double-deletion on commit bug, added test case

i-norden commented on bulk_test.go:
```diff
@@ -4,29 +4,37 @@ import (
 	"math/rand"
-	"reflect"
 	"testing"
+	// "github.com/stretchr/testify/require"
```
nit: remove this

i-norden commented on cached_mapstore.go:
```diff
@@ -0,0 +1,68 @@
+	cache   map[string]operation // map of key to operation
+	deletes map[string]struct{}
+	db      MapStore
+	cap     uint64 // starting capacity of cache
```
Did setting this starting capacity improve performance at all?

roysc replied:
Marginally - like 5% on update/insert when set to `nvalues`/10. Worth using IMO.

i-norden commented on fuzz/fuzz.go:
```diff
@@ -60,7 +58,6 @@ const (
 	Update
 	Delete
 	Prove
-	Has
```
No more Has() on the smt?

roysc replied:
Yeah, it just seemed superfluous, so I dropped it. Can always just check `value != nil`

i-norden reviewed (approved):
LGTM!!

roysc committed:
clean up

i-norden reviewed (approved):
Thanks for making the changes, LGTM!
